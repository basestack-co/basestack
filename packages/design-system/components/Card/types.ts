import { ReactNode } from "react";
import {
  LayoutProps,
  PositionProps,
  SpaceProps,
  ColorProps,
} from "styled-system";

export enum Variant {
  DEFAULT = "default",
  DANGER = "danger",
  PRIMARY = "primary",
  WARNING = "warning",
  SUCCESS = "success",
}

export interface CardProps
  extends SpaceProps,
    PositionProps,
    LayoutProps,
    ColorProps {
  /**
   * Content
   */
  children: ReactNode;
  /**
   * TestID
   */
  testId?: string;
  /**
   * Optional hover animation
   */
  hasHoverAnimation?: boolean;
  /**
   * changes styles for ex Outline color
   */
  variant?: Variant;
}
