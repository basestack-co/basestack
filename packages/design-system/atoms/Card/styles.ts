import styled from "styled-components";
import { space } from "styled-system";
import { rem } from "polished";

export const Container = styled.div`
  ${space};
  display: flex;
  flex-direction: column;
  box-shadow: ${({ theme }) => theme.shadow.elevation2};
  background-color: ${({ theme }) => theme.colors.white};
  border-radius: ${rem("6px")};
`;
