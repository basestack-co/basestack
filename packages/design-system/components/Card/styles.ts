import { rem } from "polished";
import styled, { css, type DefaultTheme } from "styled-components";
import { color, compose, layout, position, space } from "styled-system";
import { Variant } from "./types";

const getOutlineColor = (theme: DefaultTheme, variant: Variant) => {
  const colors = {
    [Variant.PRIMARY]: theme.card.variant.primary.border,
    [Variant.DANGER]: theme.card.variant.danger.border,
    [Variant.SUCCESS]: theme.card.variant.success.border,
    [Variant.WARNING]: theme.card.variant.warning.border,
    [Variant.DEFAULT]: "transparent",
  };

  return colors[variant];
};

export const Container = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "hasHoverAnimation",
})<{
  hasHoverAnimation: boolean;
  variant: Variant;
}>`
  display: flex;
  flex-direction: column;
  box-shadow: ${({ theme }) => theme.shadow.elevation2};
  background-color: ${({ theme }) => theme.card.backgroundColor};
  border-radius: ${rem("6px")};
  ${compose(space, position, layout, color)};

  ${({ hasHoverAnimation }) =>
    hasHoverAnimation &&
    css`
      transition: box-shadow 0.2s ease-in-out;

      &:hover {
        box-shadow: ${({ theme }) => theme.shadow.elevation4};
      }
    `};

  ${({ variant, theme }) =>
    variant !== Variant.DEFAULT &&
    css`
      outline-offset: -1px;
      outline: 1px solid ${getOutlineColor(theme, variant)};
    `}
`;
