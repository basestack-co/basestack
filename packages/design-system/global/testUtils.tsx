import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
