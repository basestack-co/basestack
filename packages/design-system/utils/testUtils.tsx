import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
