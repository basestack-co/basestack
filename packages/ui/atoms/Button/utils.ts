import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
