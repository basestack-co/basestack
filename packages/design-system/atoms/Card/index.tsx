import React, { memo, ReactNode } from "react";
import { SpaceProps } from "styled-system";
import { Container } from "./styles";

interface CardProps extends SpaceProps {
  /**
   * Content
   */
  children: ReactNode;
  /**
   * TestID
   */
  testId?: string;
}

const Card = ({ children, testId = "card", ...props }: CardProps) => (
  <Container data-testid={testId} {...props}>
    {children}
  </Container>
);

export default memo(Card);
