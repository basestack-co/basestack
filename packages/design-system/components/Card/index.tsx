import { forwardRef, memo } from "react";
import { Container } from "./styles";
import { CardProps, Variant } from "./types";

const Card = forwardRef<HTMLDivElement, CardProps>(
  (
    {
      variant = Variant.DEFAULT,
      children,
      testId = "card",
      hasHoverAnimation = false,
      color,
      ...props
    },
    ref,
  ) => (
    <Container
      hasHoverAnimation={hasHoverAnimation}
      ref={ref}
      data-testid={testId}
      variant={variant}
      color={color as string}
      {...props}
    >
      {children}
    </Container>
  ),
);

Card.displayName = "Card";

export { type CardProps, Variant as CardVariant };

export default memo(Card);
