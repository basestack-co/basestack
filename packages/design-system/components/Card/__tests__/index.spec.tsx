import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Card from "..";

describe("Card tests", () => {
  afterEach(cleanup);

  test("should render Card correctly", () => {
    const { asFragment } = renderWithTheme(<Card>Content</Card>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render Card with style system utils", () => {
    const { asFragment } = renderWithTheme(<Card mb={20}>Content</Card>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render Card with hover animation", () => {
    const { asFragment } = renderWithTheme(
      <Card hasHoverAnimation>Content</Card>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
