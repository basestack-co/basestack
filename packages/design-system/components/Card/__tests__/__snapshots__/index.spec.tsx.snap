// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card tests should render Card correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
}

<div
    class="c0"
    data-testid="card"
    variant="default"
  >
    Content
  </div>
</DocumentFragment>
`;

exports[`Card tests should render Card with hover animation 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
  transition: box-shadow 0.2s ease-in-out;
}

.c0:hover {
  box-shadow: rgba(0, 0, 0, 0.1) 0px 10px 15px -3px,rgba(0, 0, 0, 0.05) 0px 4px 6px -2px;
}

<div
    class="c0"
    data-testid="card"
    variant="default"
  >
    Content
  </div>
</DocumentFragment>
`;

exports[`Card tests should render Card with outline 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
  outline-offset: -1px;
  outline: 1px solid #276EF1;
}

<div
    class="c0"
    data-testid="card"
    variant="primary"
  >
    Content
  </div>
</DocumentFragment>
`;

exports[`Card tests should render Card with style system utils 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
  margin-bottom: 20px;
}

<div
    class="c0"
    data-testid="card"
    mb="20"
    variant="default"
  >
    Content
  </div>
</DocumentFragment>
`;
